assembly/src/assembler/mast_forest_builder.rs

```diff
@@ -347,79 +346,68 @@ impl MastForestBuilder {
 impl MastForestBuilder {
     /// Adds a node to the forest, and returns the [`MastNodeId`] associated with it.
     ///
-    /// If a [`MastNode`] which is equal to the current node was previously added, the previously
-    /// returned [`MastNodeId`] will be returned. This enforces this invariant that equal
-    /// [`MastNode`]s have equal [`MastNodeId`]s.
-    fn ensure_node(&mut self, node: MastNode) -> Result<MastNodeId, AssemblyError> {
-        let node_digest = node.digest();
-
-        if let Some(node_id) = self.node_id_by_hash.get(&node_digest) {
-            // node already exists in the forest; return previously assigned id
-            Ok(*node_id)
-        } else {
-            let new_node_id = self.mast_forest.add_node(node)?;
-            self.node_id_by_hash.insert(node_digest, new_node_id);
-
-            Ok(new_node_id)
-        }
+    /// Note adding the same [`MastNode`] twice will result in two different [`MastNodeId`]s being
+    /// returned.
+    pub fn add_node(&mut self, node: MastNode) -> Result<MastNodeId, AssemblyError> {
+        Ok(self.mast_forest.add_node(node)?)
     }
 
     /// Adds a basic block node to the forest, and returns the [`MastNodeId`] associated with it.
-    pub fn ensure_block(
+    pub fn add_block(
         &mut self,
         operations: Vec<Operation>,
         decorators: Option<DecoratorList>,
     ) -> Result<MastNodeId, AssemblyError> {
         let block = MastNode::new_basic_block(operations, decorators)?;
-        self.ensure_node(block)
+        self.add_node(block)
     }
```

Comment:
This would create a new node for every basic block, right?

I was thinking we'd create new nodes only if either MAST roots or decorators are different - i.e., if MAST roots and decorators match - this would behave like it did before: return a reference to an already existing node.'

So, maybe this can remain `ensure_node()` just the internal representation would change.

One implication of this is that for every node we'd define hash as `hash(mast_root || decorator_hash)` and then use this hash to lookup nodes instead of MAST root as previously.

Reply:
Didn't we want to avoid computing 2 separate hashes for performance reasons?

Also, we now have a new invariant that procedures are uniquely identified by their `MastNodeId`, so in the case of procedure roots, we need to force a new `MastNodeId` no matter what. In which case we could expose `ensure_x()` and `add_x()`

Reply:
> Didn't we want to avoid computing 2 separate hashes for performance reasons?

This is only during assembly - so, some performance degradation should be OK. Also, since decorator hash impacts only how we add nodes to the MAST forest, and can use a fast hash function for this (i.e., BLAKE3) - and so performance impact should be minimal).

> Also, we now have a new invariant that procedures are uniquely identified by their `MastNodeId`, so in the case of procedure roots, we need to force a new `MastNodeId` no matter what.

Why do we need this invariant? If two procedures have the same set of operations and decorators, could we not have them be the same `MastNode`?

Reply:
Hmm, right, I was using this invariant to ensure that we always invoke the right procedure (when identical MAST roots but different decorators), but it's stronger than it needs to be.

Reply:
fixed